src/parser/input_utils.rs

```diff
@@ -0,0 +1,98 @@
+use crate::parser::utils::{string_to_node, NixNode};
+use anyhow::{anyhow, bail};
+use either::Either;
+use rnix::{types::*, NixLanguage, StrPart, SyntaxKind::*, parse};
+use rowan::{api::SyntaxNode, GreenNode, GreenNodeBuilder, Language, NodeOrToken, GreenToken};
+use std::string::ToString;
+use nixpkgs_fmt::reformat_node;
+use std::ops::Deref;
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+struct Input {
+    url: String,
+    is_flake: bool,
+}
+
+pub fn wrap_root(node: GreenNode) -> NixNode {
+    let kind: rowan::SyntaxKind = NixLanguage::kind_to_raw(NODE_ROOT);
+    let root = GreenNode::new(kind, (vec![NodeOrToken::Node(node)]));
+    SyntaxNode::new_root(root)
+}
+
+pub fn new_string(s : String) -> GreenNode {
+    let mut node = GreenNodeBuilder::new();
+    let kind: rowan::SyntaxKind = NixLanguage::kind_to_raw(NODE_STRING);
+    node.start_node(kind);
+    let start_string_kind: rowan::SyntaxKind = NixLanguage::kind_to_raw(TOKEN_STRING_START);
+    node.token(start_string_kind, "\"");
+    let string_content: rowan::SyntaxKind = NixLanguage::kind_to_raw(TOKEN_STRING_CONTENT);
+    node.token(string_content, &s);
+    let end_string_kind: rowan::SyntaxKind = NixLanguage::kind_to_raw(TOKEN_STRING_END);
+    node.token(end_string_kind, "\"");
+    node.finish_node();
+    node.finish()
+}
+
+pub fn new_key(s : String) -> GreenNode {
+    let kind: rowan::SyntaxKind = NixLanguage::kind_to_raw(NODE_KEY);
+    let children = vec![NodeOrToken::Node(new_string(s))];
+    GreenNode::new(kind, children)
+}
+
+pub fn gen_key_value(key: String, value: String) -> GreenNode {
+    let key_node : GreenNode = new_key(key);
+    let value_node : GreenNode = new_string(value);
+    new_key_value(key_node, value_node)
+}
+
+pub fn new_key_value(key : GreenNode, value: GreenNode) -> GreenNode {
+    let kind = NixLanguage::kind_to_raw(NODE_KEY_VALUE);
+    let assign_kind = NixLanguage::kind_to_raw(TOKEN_ASSIGN);
+    let whitespace_kind = NixLanguage::kind_to_raw(TOKEN_WHITESPACE);
+    let semicolon_kind = NixLanguage::kind_to_raw(TOKEN_SEMICOLON);
+    let children = vec![
+        NodeOrToken::Node(key),
+        NodeOrToken::Token(GreenToken::new(whitespace_kind, " ")),
+        NodeOrToken::Token(GreenToken::new(assign_kind, "=")),
+        NodeOrToken::Token(GreenToken::new(whitespace_kind, " ")),
+        NodeOrToken::Node(value),
+        NodeOrToken::Token(GreenToken::new(semicolon_kind, ";")),
+    ];
+    GreenNode::new(kind, children)
+}
+
+pub fn gen_attr_set(attr_pairs: Vec<(String, String)>) -> GreenNode{
+    let new_attr_pairs : Vec<(GreenNode, GreenNode)> =
+        attr_pairs.iter().map(|(key, value)| (new_key(key.to_string()), new_string(value.to_string()))).collect();
+    new_attr_set(new_attr_pairs)
```

Comment:
from a performance standpoint it is a better idea to just merge `gen_attr_set` and `new_attr_set`, to avoid the allocation of `new_attr_pairs`, which then get `.clone()`d directly again, which should be avoided.

Reply:
What I ended up doing here is implementing the from trait so I can just convert directly from an Input to a GreenNode. The `gen_*` functions can be deleted.